spyder/widgets/sourcecode/base.py

```diff
@@ -317,43 +319,61 @@ def set_palette(self, background, foreground):
 
 
     #------Extra selections
-    def extra_selection_length(self, key):
-        selection = self.get_extra_selections(key)
-        if selection:
-            cursor = self.extra_selections_dict[key][0].cursor
-            selection_length = cursor.selectionEnd() - cursor.selectionStart()
-            return selection_length
-        else:
-            return 0
-
     def get_extra_selections(self, key):
+        """Return editor extra selections.
+
+        Args:
+            key (str) name of the extra selections group
+
+        Returns:
+            list of sourcecode.api.TextDecoration.
+        """
         return self.extra_selections_dict.get(key, [])
 
     def set_extra_selections(self, key, extra_selections):
+        """Set extra selections for a key.
+
+        Also assign draw orders to leave current_cell and current_line
+        in the backgrund (and avoid them to cover other decorations)
+
+        NOTE: This will remove previous decorations added to  the same key.
+
+        Args:
+            key (str) name of the extra selections group.
+            extra_selections (list of sourcecode.api.TextDecoration).
+        """
+        # current line has to be highlighted first
+        draw_orders = {'current_cell': 0,
+                       # draw order = 1 is used in codefolding
+                       'current_line': 2}
+        draw_order = draw_orders.get(key,  5)

```

Comment:
Why `5` here?

Reply:
It just have to be a number greater that current_line and curreent_cell.

I choose 5 to leave 3 and 4 free, maybe somebody will need to add something in front of the current line, but behind if the other selections.

Reply:
Please define and use constants for this. Something like:

    CELL_ORDER = 0
    FOLDING_ODER = 1
    LINE_ORDER = 2

and use those constants here.

Reply:
I'll use this same dict an expose it as a part of the decoration api.

Reply:
Great!

Reply:
I just add a commit with the DRAW_ORDERS dict, I didn't leave 3 and 4 unused because now its easier to modify the draworders values.